### PR TITLE
IO: Refactor directory layout

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #8691

Problem Summary:

### What is changed and how it works?

This PR seems huge, but it is quite simple actually.

1. Refactor `IO/` directory layout
    a. move buffer related files under `IO/Buffer`
    b. move FileProvider related files under `IO/`
    b. move `Encryption/` under `IO/`
2. remove legacy test code which are not used anymore.

713e7846-3e8e-49e1-83b6-fbe741a13228

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
```
$ tree dbms/src/IO/
dbms/src/IO/
├── BaseFile
│   ├── fwd.h
│   ├── IORateLimitConfig.cpp
│   ├── IORateLimitConfig.h
│   ├── MemoryRandomAccessFile.h
│   ├── PosixRandomAccessFile.cpp
│   ├── PosixRandomAccessFile.h
│   ├── PosixWritableFile.cpp
│   ├── PosixWritableFile.h
│   ├── PosixWriteReadableFile.cpp
│   ├── PosixWriteReadableFile.h
│   ├── RandomAccessFile.h
│   ├── RateLimiter.cpp
│   ├── RateLimiter.h
│   ├── README.md
│   ├── tests
│   │   └── gtest_rate_limiter.cpp
│   ├── WritableFile.h
│   └── WriteReadableFile.h
├── Buffer
│   ├── AsynchronousWriteBuffer.h
│   ├── BufferBase.h
│   ├── BufferWithOwnMemory.h
│   ├── ConcatReadBuffer.h
│   ├── createReadBufferFromFileBase.cpp
│   ├── createReadBufferFromFileBase.h
│   ├── createWriteBufferFromFileBase.cpp
│   ├── createWriteBufferFromFileBase.h
│   ├── HexWriteBuffer.cpp
│   ├── HexWriteBuffer.h
│   ├── IReadableWriteBuffer.h
│   ├── LimitReadBuffer.cpp
│   ├── LimitReadBuffer.h
│   ├── LinearMemoryWriteBuffer.cpp
│   ├── LinearMemoryWriteBuffer.h
│   ├── MemoryReadWriteBuffer.cpp
│   ├── MemoryReadWriteBuffer.h
│   ├── ReadBufferAIO.cpp
│   ├── ReadBufferAIO.h
│   ├── ReadBuffer.cpp
│   ├── ReadBufferFromFileBase.cpp
│   ├── ReadBufferFromFileBase.h
│   ├── ReadBufferFromFile.cpp
│   ├── ReadBufferFromFileDescriptor.cpp
│   ├── ReadBufferFromFileDescriptor.h
│   ├── ReadBufferFromFile.h
│   ├── ReadBufferFromIStream.h
│   ├── ReadBufferFromMemory.h
│   ├── ReadBufferFromPocoSocket.cpp
│   ├── ReadBufferFromPocoSocket.h
│   ├── ReadBufferFromRandomAccessFile.cpp
│   ├── ReadBufferFromRandomAccessFile.h
│   ├── ReadBufferFromString.h
│   ├── ReadBuffer.h
│   ├── StdStreamBufFromReadBuffer.cpp
│   ├── StdStreamBufFromReadBuffer.h
│   ├── StdStreamFromReadBuffer.h
│   ├── tests
│   │   ├── gtest_ios_wrapper.cpp
│   │   ├── gtest_std_stream_from_read_buffer.cpp
│   │   └── gtest_write_buffer_from_string.cpp
│   ├── WriteBufferAIO.cpp
│   ├── WriteBufferAIO.h
│   ├── WriteBufferFromArena.h
│   ├── WriteBufferFromFileBase.cpp
│   ├── WriteBufferFromFileBase.h
│   ├── WriteBufferFromFile.cpp
│   ├── WriteBufferFromFileDescriptor.cpp
│   ├── WriteBufferFromFileDescriptor.h
│   ├── WriteBufferFromFile.h
│   ├── WriteBufferFromOStream.h
│   ├── WriteBufferFromPocoSocket.cpp
│   ├── WriteBufferFromPocoSocket.h
│   ├── WriteBufferFromString.h
│   ├── WriteBufferFromTemporaryFile.cpp
│   ├── WriteBufferFromTemporaryFile.h
│   ├── WriteBufferFromVector.h
│   ├── WriteBufferFromWritableFile.cpp
│   ├── WriteBufferFromWritableFile.h
│   ├── WriteBuffer.h
│   ├── WriteBufferValidUTF8.cpp
│   └── WriteBufferValidUTF8.h
├── Checksum
│   ├── ChecksumBuffer.cpp
│   ├── ChecksumBuffer.h
│   ├── ChecksumReadBufferBuilder.cpp
│   ├── ChecksumReadBufferBuilder.h
│   ├── ChecksumWriteBufferBuilder.cpp
│   ├── ChecksumWriteBufferBuilder.h
│   └── tests
│       └── gtest_dm_checksum_buffer.cpp
├── CMakeLists.txt
├── Compression
│   ├── CodecDeflateQpl.cpp
│   ├── CodecDeflateQpl.h
│   ├── CompressedReadBufferBase.cpp
│   ├── CompressedReadBufferBase.h
│   ├── CompressedReadBuffer.cpp
│   ├── CompressedReadBufferFromFile.cpp
│   ├── CompressedReadBufferFromFile.h
│   ├── CompressedReadBuffer.h
│   ├── CompressedStream.h
│   ├── CompressedWriteBuffer.cpp
│   ├── CompressedWriteBuffer.h
│   ├── CompressionSettings.cpp
│   └── CompressionSettings.h
├── ConnectionTimeouts.h
├── copyData.cpp
├── copyData.h
├── DoubleConverter.h
├── Encryption
│   ├── AESCTRCipher.cpp
│   ├── AESCTRCipher.h
│   ├── AESCTRCipherStream.cpp
│   ├── AESCTRCipherStream.h
│   ├── BlockAccessCipherStream.h
│   ├── CMakeLists.txt
│   ├── CompressedReadBufferFromFileProvider.cpp
│   ├── CompressedReadBufferFromFileProvider.h
│   ├── DataKeyManager.cpp
│   ├── DataKeyManager.h
│   ├── EncryptedRandomAccessFile.cpp
│   ├── EncryptedRandomAccessFile.h
│   ├── EncryptedWritableFile.cpp
│   ├── EncryptedWritableFile.h
│   ├── EncryptedWriteReadableFile.cpp
│   ├── EncryptedWriteReadableFile.h
│   ├── EncryptionKey.h
│   ├── KeyspacesKeyManager.cpp
│   ├── KeyspacesKeyManager.h
│   ├── MasterKey.h
│   ├── MockKeyManager.cpp
│   ├── MockKeyManager.h
│   └── tests
│       ├── gtest_encryption_test.cpp
│       └── gtest_keyspaces_key_manager.cpp
├── EncryptionPath.h
├── Endian.h
├── FileProvider.cpp
├── FileProvider_fwd.h
├── FileProvider.h
├── HTTPCommon.cpp
├── HTTPCommon.h
├── HTTPHeaderEntries.h
├── IOSWrapper.h
├── IOThreadPool.h
├── IOThreadPools.h
├── KeyManager.h
├── Operators.h
├── parseDateTimeBestEffort.cpp
├── parseDateTimeBestEffort.h
├── Progress.cpp
├── Progress.h
├── ReadBufferFromRandomAccessFileBuilder.cpp
├── ReadBufferFromRandomAccessFileBuilder.h
├── readFloatText.cpp
├── readFloatText.h
├── ReadHelpers.cpp
├── ReadHelpers.h
├── UseSSL.cpp
├── UseSSL.h
├── VarInt.h
├── WriteBufferFromWritableFileBuilder.cpp
├── WriteBufferFromWritableFileBuilder.h
├── WriteHelpers.cpp
├── WriteHelpers.h
└── WriteIntText.h

9 directories, 158 files
```

File: dbms/CMakeLists.txt [hunk 1]
```diff
@@ -72,7 +72,10 @@ add_headers_and_sources(dbms src/Debug)
 add_headers_and_sources(dbms src/Debug/MockExecutor)
 add_headers_and_sources(dbms src/Debug/MockKVStore)
 add_headers_and_sources(dbms src/Dictionaries)
-add_headers_and_sources(dbms src/Encryption)
+# Encryption needs to link tiflash-proxy, but we do not want to link tiflash-proxy to tiflash_common_io.
+# So we put the encryption source files in dbms.
+add_headers_and_sources(dbms src/IO)
+add_headers_and_sources(dbms src/IO/Encryption)
```
Comment (JaySon-Huang) [inline dbms/CMakeLists.txt:78]: Can we keep it as following ?
```
add_headers_and_sources(tiflash_common_io src/IO)
....
add_headers_and_sources(dbms src/IO/Encryption)
```

File: dbms/src/Common/tests/bench_logger.cpp
Comment (JaySon-Huang) [inline dbms/src/Common/tests/bench_logger.cpp:1]: keep this file

File: dbms/CMakeLists.txt (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/CMakeLists.txt:78]: no, FileProvider rely on `Encryption/`

**Review by JaySon-Huang — APPROVED** [review]

**Review by CalvinNeo — APPROVED** [review]
lgmt

**Pull request merged**